File: tools/_c252-liste.mjs
// C-252: die Katalogliste GEGEN DIE LAUFENDE DATENBANK zaehlen.
//
// `[read]` Typecheck sagt nichts ueber PostgREST — eine falsche
// Einbettung faellt erst zur Laufzeit auf. Deshalb wird hier dieselbe
// Abfrage gestellt, die `ladeSubstanzListe` stellt.
import { createClient } from '@supabase/supabase-js'
import { readFileSync } from 'node:fs'

const env = Object.fromEntries(
  readFileSync('apps/web/.env.local', 'utf8')
    .split('\n')
    .filter(z => z.includes('=') && !z.trim().startsWith('#'))
    .map(z => {
      const i = z.indexOf('=')
      return [z.slice(0, i).trim(), z.slice(i + 1).trim().replace(/^["']|["']$/g, '')]
    }))

const url = env.NEXT_PUBLIC_SUPABASE_URL
const key = env.SUPABASE_SERVICE_ROLE_KEY ?? env.SUPABASE_SERVICE_ROLE
const s = createClient(url, key, { auth: { persistSession: false } }).schema('supplements')

const AUSWAHL = 'id, slug, name_de, name_en, description_de, description_en, form,'
  + ' evidence_grade, source,'
  + ' supplement_categories(name_de, name_en),'
  + ' supplement_groups(code, label_de, label_en)'

// ── MIT Filter: die Erwartung ist 290 ──────────────────────────────
const mit = await s.from('supplements').select(AUSWAHL, { count: 'exact' })
  .eq('im_katalog', true).order('sort_order').order('name_en')
if (mit.error) { console.error('FEHLER (mit Filter):', mit.error.message); process.exit(1) }
console.log(`MIT  .eq('im_katalog', true) : ${mit.count} Eintraege   (erwartet 290)`)

// ── OHNE Filter: die Negativprobe, Erwartung 566 ───────────────────
const ohne = await s.from('supplements').select(AUSWAHL, { count: 'exact' })
  .order('sort_order').order('name_en')
if (ohne.error) { console.error('FEHLER (ohne Filter):', ohne.error.message); process.exit(1) }
console.log(`OHNE Filter                  : ${ohne.count} Eintraege   (erwartet 566)`)

// ── Die Gegenprobe, namentlich ─────────────────────────────────────
const namen = new Set((mit.data ?? []).map(r => r.slug))
console.log()
console.log('Gegenprobe, namentlich:')
console.log(`  sub_9f9bb8c160 (Creatine monohydrate, Grad A) in der Liste: `
  + `${namen.has('sub_9f9bb8c160') ? 'JA  (richtig)' : 'NEIN (FEHLER)'}`)
console.log(`  f05_7_keto_dhea (7-Keto DHEA, ohne Beschreibung)  in der Liste: `
  + `${namen.has('f05_7_keto_dhea') ? 'JA  (FEHLER)' : 'NEIN (richtig)'}`)

// ── Kommen die eingebetteten Tabellen wirklich an? ─────────────────
const p = (mit.data ?? []).find(r => r.slug === 'sub_9f9bb8c160')
console.log()
console.log('Ein Eintrag, wie ihn die Liste liefert:')
console.log(`  name_de=${JSON.stringify(p?.name_de)}  name_en=${JSON.stringify(p?.name_en)}`)
console.log(`  Kategorie=${JSON.stringify(p?.supplement_categories)}`)
console.log(`  Gruppe=${JSON.stringify(p?.supplement_groups)}`)
console.log(`  Grad=${p?.evidence_grade}  form=${JSON.stringify(p?.form)}`)

// ── Wieviele traegen nach der Umsetzung wirklich einen Namen? ──────
const ohneName = (mit.data ?? []).filter(r => !r.name_de && !r.name_en).length
const ohneBeschr = (mit.data ?? []).filter(r => !r.description_de && !r.description_en).length
const ohneGruppe = (mit.data ?? []).filter(r => !r.supplement_groups?.code).length
console.log()
console.log(`Von den ${mit.count}: ohne Namen ${ohneName} · ohne Beschreibung `
  + `${ohneBeschr} · ohne Gruppe ${ohneGruppe}   (jeweils erwartet 0)`)

File: tools/_c252-liste.py
"""C-252: die Katalogliste gegen die laufende Datenbank zaehlen.

`[read]` Geprueft wird die SQL-Wirkung der Abfrage, die
`ladeSubstanzListe` stellt — mit und ohne `im_katalog`-Filter, plus die
namentliche Gegenprobe. Ueber `psql`, nicht ueber den Supabase-Client:
der braeuchte den Service-Key aus `.env.local`, und den zu lesen sperrt
der Hook zu Recht.
"""
import sys
sys.path.insert(0, r"D:\GitHub\LumeOS-Claude-V1\tools")
from lauf import psql

print("=== Die Zahl, um die es geht ===")
print(psql("""select
  count(*) filter (where im_katalog) as mit_filter,
  count(*) as ohne_filter
  from supplements.supplements;"""))
print("   erwartet: mit_filter 290 · ohne_filter 566")

print()
print("=== Gegenprobe, namentlich ===")
print(psql("""select slug, name_en, im_katalog,
    coalesce(evidence_grade,'(kein Grad)') as grad,
    case when coalesce(description_en,'')='' then '(keine)' else 'vorhanden' end as beschreibung
  from supplements.supplements
  where slug in ('sub_9f9bb8c160','f05_7_keto_dhea')
  order by im_katalog desc;"""))
print("   sub_9f9bb8c160 muss erscheinen (im_katalog t)")
print("   f05_7_keto_dhea darf nicht erscheinen (im_katalog f)")

print()
print("=== Traegt jeder der 290 die Felder, die die Liste zeigt? ===")
print(psql("""select
  count(*) as gesamt,
  count(*) filter (where coalesce(name_de, name_en, '')='') as ohne_namen,
  count(*) filter (where coalesce(description_de, description_en, '')='') as ohne_beschreibung,
  count(*) filter (where group_id is null) as ohne_gruppe,
  count(*) filter (where category_id is null) as ohne_kategorie,
  count(*) filter (where evidence_grade is null) as ohne_grad
  from supplements.supplements where im_katalog;"""))
print("   erwartet: 290 und danach ueberall 0")

print()
print("=== Die Einbettung: kommt Kategorie und Gruppe an? ===")
print(psql("""select s.slug,
    coalesce(nullif(s.name_de,''), s.name_en) as name,
    coalesce(nullif(k.name_de,''), k.name_en) as kategorie,
    g.code as gruppe, s.evidence_grade as grad
  from supplements.supplements s
  left join supplements.supplement_categories k on k.id = s.category_id
  left join supplements.supplement_groups g on g.id = s.group_id
  where s.slug = 'sub_9f9bb8c160';"""))

print()
print("=== Die Gruppen, wie die Liste sie zaehlt (nur die 290) ===")
print(psql("""select g.code, count(*)
  from supplements.supplements s
  join supplements.supplement_groups g on g.id = s.group_id
  where s.im_katalog group by 1 order by 2 desc;"""))
